agent/xds/delta.go

```diff
@@ -0,0 +1,728 @@
+		if version == "" {
+			delete(t.resourceVersions, name)
+		} else {
+			t.resourceVersions[name] = version
+		}
+	}
+	t.sentToEnvoyOnce = true
+	delete(t.pendingUpdates, nonce)
+}
+
+func (t *xDSDeltaType) nack(nonce string) {
+	delete(t.pendingUpdates, nonce)
+}
+
+func (t *xDSDeltaType) SendIfNew(
+	kind structs.ServiceKind,
+	currentVersions map[string]string, // type => name => version (as consul knows right now)
+	resourceMap IndexedResources,
+	nonce *uint64,
+	upsert, remove bool,
+) (error, bool) {
+	if t == nil || !t.registered {
+		return nil, false
+	}
+	logger := t.generator.Logger.With("typeUrl", t.typeURL)
+
+	allowEmpty := t.allowEmptyFn != nil && t.allowEmptyFn(kind)
+
+	// Zero length resource responses should be ignored and are the result of no
+	// data yet. Notice that this caused a bug originally where we had zero
+	// healthy endpoints for an upstream that would cause Envoy to hang waiting
+	// for the EDS response. This is fixed though by ensuring we send an explicit
+	// empty LoadAssignment resource for the cluster rather than allowing junky
+	// empty resources.
+	if len(currentVersions) == 0 && !allowEmpty {
+		// Nothing to send yet
+		return nil, false
+	}
+
+	resp, updates, err := t.createDeltaResponse(currentVersions, resourceMap, upsert, remove)
+	if err != nil {
+		return err, false
+	}
+
+	if resp == nil {
+		return nil, false
+	}
+
+	*nonce++
+	resp.Nonce = fmt.Sprintf("%08x", *nonce)
+
+	t.generator.logTraceResponse("Incremental xDS v3", resp)
+
+	logger.Trace("sending response", "nonce", resp.Nonce)
+	if err := t.stream.Send(resp); err != nil {
+		return err, false
+	}
+	logger.Trace("sent response", "nonce", resp.Nonce)
+
+	t.pendingUpdates[resp.Nonce] = updates
+
+	return nil, true
+}
+
+func (t *xDSDeltaType) createDeltaResponse(
+	currentVersions map[string]string, // name => version (as consul knows right now)
+	resourceMap IndexedResources,
+	upsert, remove bool,
+) (*envoy_discovery_v3.DeltaDiscoveryResponse, map[string]string, error) {
+	// compute difference
+	var (
+		hasRelevantUpdates = false
+		updates            = make(map[string]string)
+	)
+	// First find things that need updating or deleting
+	for name, envoyVers := range t.resourceVersions {
+		currVers, ok := currentVersions[name]
+		if !ok {
+			if remove {
+				hasRelevantUpdates = true
+			}
+			updates[name] = ""
+		} else if currVers != envoyVers {
+			if upsert {
+				hasRelevantUpdates = true
+			}
+			updates[name] = currVers
+		}
+	}
+
+	// Now find new things
+	if t.wildcard {
+		for name, currVers := range currentVersions {
+			if _, ok := t.resourceVersions[name]; !ok {
+				updates[name] = currVers
+				if upsert {
+					hasRelevantUpdates = true
+				}
+			}
+		}
+	}
+
+	if !hasRelevantUpdates && t.sentToEnvoyOnce {
+		return nil, nil, nil
+	}
+
+	// now turn this into a disco response
+	resp := &envoy_discovery_v3.DeltaDiscoveryResponse{
+		// TODO(rb): consider putting something in SystemVersionInfo?
```

Comment:
If we send some string in `SystemVersionInfo` envoy will log it. This can help follow along over on the envoy side. When I was debugging something here I stuffed much of the state of the `xDSDeltaType` structure here which helped. Since our updates aren't officially versioned but versioned as the snapshot evolves there really isn't a useful human-readable string I could come up with to pass in general.